sail/src/main/java/org/apache/rya/rdftriplestore/inference/InferenceEngine.java

```diff
@@ -762,6 +800,57 @@ public void handleStatement(final Statement statement) throws RDFHandlerExceptio
     }
 
     /**
+     * For a given type, return any properties and values such that owl:hasSelf restrictions on
+     * those properties could imply this type. No matter how many restrictions are returned, each
+     * one is considered individually sufficient: if a resource has the property and the value, then
+     * it belongs to the provided type. Takes type hierarchy into account, so the value may imply a
+     * subtype which in turn implies the provided type.
+     * @param type The type (URI or bnode) to check against the known restrictions
+     * @return For each relevant property, a set of values such that whenever a resource has that
+     *      value for that property, it is implied to belong to the type.
+     */
+    public Set<URI> getHasSelfImplyingType(final Resource type){
+        // return properties that imply this type if reflexive
+        final Set<URI> properties = new HashSet<>();
+        properties.addAll(hasSelfByType.get(type));
+        //findParent gets all subclasses, add self.
+        if (type instanceof URI) {
+            for (final URI subtype : findParents(subClassOfGraph, (URI) type)) {
+                properties.addAll(hasSelfByType.get(subtype));
+            }
+        }
+
+        // make map hasSelfByType[]
+        return properties;
+    }
+
+    /**
+     * For a given property, return any types such that some owl:hasSelf restriction implies that members
+     * of the type have the value of themselves for this property.
+     *
+     * This takes into account type hierarchy, where children of a type that have
+     * this property are also assumed to have the property.
+     * @param property The property whose owl:hasSelf restrictions to return
+     * @return A set of types that possess the implied property.
+     */
+    public Set<Resource> getHasSelfImplyingProperty(final URI property) {
+        // return types that imply this type if reflexive
+        final Set<Resource> types = new HashSet<>();
+        final Set<Resource> baseTypes = hasSelfByProperty.get(property);
```

Comment:
Same issue as with getHasSelfImplyingType -- check that baseTypes isn't null before calling addAll and iterating through it